### PR TITLE
Read file URL in chunks

### LINKED ISSUE #21466: Read file URLs in chunks
The code to [read from file URLs](https://github.com/servo/servo/blob/master/components/net/fetch/methods.rs#L483-L506) uses read_to_end to consume the entire file buffer and send it back all at once. This means the behaviour of HTTP URLs and file URLs for large files is very different.

### PR DESCRIPTION
**Pull request by snowsignal**
This is a very straightforward PR: essentially, I replaced a `read_to_end` call that occurs when processing a file URL fetch with a loop that reads the file in chunks (with a `FILE_CHUNK_SIZE` constant that specifies the chunk size, which I've put at 32KB), and then calls `target.process_response_chunk` to process the chunk. The chunk is then appended to the fetch result, and once the end of the file is reached, the result is returned.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #21466.

- [ ] There are tests for these changes OR
- [X] These changes do not require tests because there isn't really a way to observe this.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/21560)

### CODE REVIEW

File: components/net/fetch/methods.rs [hunk 1]
```diff
@@ -486,8 +488,20 @@ fn scheme_fetch(request: &mut Request,
...
-                                let mut bytes = vec![];
-                                let _ = file.read_to_end(&mut bytes);
+                                let mut reader = BufReader::with_capacity(FILE_CHUNK_SIZE, file);
+                                let mut bytes = Vec::new();
+                                loop {
+                                    let length = {
+                                        let mut buffer = reader.fill_buf().unwrap().to_vec();
+                                        let buffer_len = buffer.len();
+                                        bytes.append(&mut buffer);
+                                        target.process_response_chunk(buffer);
```
Comment (jdm) [inline components/net/fetch/methods.rs]: We actually have a system in place that requires less manual work, so let's make use of it. `wait_for_response` handles both synchronous and asynchronous responses; we can make this one asynchronous by:
1) spawning a new thread before we open the file
2) creating a new channel and storing it in `done_chan` (https://github.com/servo/servo/blob/f04c965a9bab46ff6943416baab338032bc526c1/components/net/http_loader.rs#L1091-L1093)
3) setting the body to ResponseBody::Receiving
4) sending the file chunks across the channel
5) setting the body to ResponseBody::Done once the file is completely read (https://github.com/servo/servo/blob/f04c965a9bab46ff6943416baab338032bc526c1/components/net/http_loader.rs#L1127-L1151)

To ensure everything still works, try loading a page from your local machine (not from a web server) that loads an image >32kb using a relative URL.

**Review by jdm — CHANGES_REQUESTED** [review]

File: components/net/fetch/methods.rs (hunk 1 above)
Comment (JacksonCoder) [reply components/net/fetch/methods.rs]: Would `thread::spawn` work for making a new thread?

Comment (jdm) [reply components/net/fetch/methods.rs]: Yes.

Comment (JacksonCoder) [reply components/net/fetch/methods.rs]: Should I still call `target.process_response_chunk`?

Comment (jdm) [reply components/net/fetch/methods.rs]: That should not be necessary in the model that I described.

Comment (JacksonCoder) [reply components/net/fetch/methods.rs]: I had some time to work on this today, I'm building it to see if I can load files properly. I'll push the commit if it works.

**Review by jdm — APPROVED** [review]

**Comment by jdm** [comment]
```
Checking files for tidiness...
./components/net/fetch/methods.rs:31: use statement is not in alphabetical order
	expected: std::sync::mpsc::channel
	found: std::str
```

File: components/net/fetch/methods.rs [hunk 2]
```diff
@@ -486,13 +490,51 @@ fn scheme_fetch(request: &mut Request,
...
+                                            *res_body.lock().unwrap() = ResponseBody::Done(vec![]);
+                                            let _ = done_sender.send(Data::Cancelled);
+                                            return;
+                                        }
+                                        let length = {
+                                            let mut buffer = reader.fill_buf().unwrap().to_vec();
+                                            let buffer_len = buffer.len();
+                                            if let ResponseBody::Receiving(ref mut body) = *res_body.lock().unwrap() {
+                                                body.extend_from_slice(&buffer);
+                                                let _ = done_sender.send(Data::Payload(buffer.clone()));
```
Comment (jdm) [inline components/net/fetch/methods.rs]: Why is this clone necessary?

**Comment by jdm** [comment]
Apart from those small nits, these changes look great! Thanks for working on this!

File: components/net/fetch/methods.rs (hunk 2 above)
Comment (JacksonCoder) [reply components/net/fetch/methods.rs]: It isn't, I forgot to remove it because I was doing some testing earlier. It's fixed now.

**Comment by JacksonCoder** [comment]
No problem, I loved working on this and hope to get more involved in Servo development in the future.

**Pull request merged**